shub/apps/library/views/images.py

```diff
@@ -551,6 +554,60 @@ def get(self, request):
         collections = generate_collections_list(token.user)
         return Response(data=collections, status=200)
 
+    def post(self, request):
+        """Create a new collection.
+
+        POST /v1/collections
+
+        Body parameters:
+        * entity: entity id (= user id)
+        * name: new collection name
+        * private: optional boolean (defaults to the configured default for
+          new collections
+
+        Return the newly created collection.
+        """
+
+        print("POST CollectionsView")
+        if not validate_token(request):
+            print("Token not valid")
+            return Response(status=403)
+
+        # body should have {'entity': entity_id, 'name': new_collection_name, 'private': bool}
+        # 'private' is optional and defaults to Collection.private default
+        # value set with get_privacy_default()
+        # {"entity": "42", "name": "my_collection", "private": true}
+        body = json.loads(request.body.decode("utf-8"))
+        if not ("entity" in body and "name" in body):
+            print("Invalid payload")
+            return Response(status=400)
+
+        # does a collection with the same name exist already?
+        if not get_collection(body["name"], False) is None:
+            print("A collection named '{0}' exists already!".format(body["name"]))
+            return Response(status=403)
+
+        # check user permission
+        token = get_token(request)
+        if str(token.user.id) != body["entity"]:
+            print("Permission denied {0} {1}".format(token.user.id, body["entity"]))
+            return Response(status=403)
+
+        # create the collection
+        name = format_collection_name(body["name"])
+        collection = Collection(name=name, secret=str(uuid.uuid4()))
```

Comment:
Do you not want to do a proper `Create` here (which would handle that first save?) And could this fail for any reason and throw an exception?

Reply:
Well, I'm not familiar at all with the django framework and didn't find such a method in the source tree. How is this `Create` supposed to be used?

Reply:
If you think something might exist, you do:

```python
instance, created = Model.objects.get_or_create(name=...)
```
But since you know this doesn't exist, you would just do:
```python
instance = Model.objects.Create(name=...)
```
It saves you the step of needing to do `instance.save()`. For the way you did it, typically that's done if you need to create the instance (without saving) because it's missing some required field.

Reply:
There is good docs starting here https://docs.djangoproject.com/en/3.1/ref/models/querysets/#create

Reply:
How about this?
```python
        name = format_collection_name(body["name"])
        try:
            collection = Collection.objects.get(name=name)
            print("A collection named '{0}' exists already!".format(name))
            return Response(status=403)
        except Collection.DoesNotExist:
            pass

        collection.secret = str(uuid.uuid4())
        collection.owners.add(token.user)

        # set privacy if present
        if "private" in body:
            collection.private = body["private"]

        collection.save()
```

Reply:
Sorry, I think you were working on the wrong thing here? This functionality is already handled above by `get_collection`. The Create that we want to handle is below that, you currently have:

```python
collection = Collection(name=name, secret=str(uuid.uuid4()))
```
And as I mentioned above, you'd want to use Create

```python
collection = Collection.objects.create(name=name, secret=str(uuid.uuid4()))
```
You've already established that it does not exist with `get_collection` before that.

Reply:
I've grouped both topics (check then create) into this snippet.